File: swift/Sources/LibSignalClient/chat/AuthPaymentsService.swift
//
// Copyright 2026 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

import Foundation

public struct CurrencyConversions: Equatable {
    public init(timestamp: Date, currencies: [Currency]) {
        self.timestamp = timestamp
        self.currencies = currencies
    }
    internal init(fromInternal cci: CurrencyConversionsInternal) {
        self.timestamp = cci.timestampMs
        self.currencies = cci.currencies.map { c in
            Currency(base: c.base, conversions: Dictionary(uniqueKeysWithValues: c.conversions))
        }
    }

    public let timestamp: Date
    public let currencies: [Currency]
}
public struct Currency: Equatable {
    public init(base: String, conversions: [String: String]) {
        self.base = base
        self.conversions = conversions
    }

    public let base: String
    /// The values of this map are decimal conversion rates.
    public let conversions: [String: String]
}

public protocol AuthPaymentsService: Sendable {
    /// Return the current currency conversion rates.
    ///
    /// - Throws:
    ///   - the standard Signal network errors
    func getCurrencyConversions() async throws -> CurrencyConversions
}

extension AuthenticatedChatConnection: AuthPaymentsService {

    public func getCurrencyConversions() async throws -> CurrencyConversions {
        return CurrencyConversions(
            fromInternal: try await NativeNice.AuthenticatedChatConnection_get_currency_conversions(
                asyncContext: self.tokioAsyncContext,
                chat: self,
            )
        )
    }

}

extension AuthServiceSelector where Self == AuthServiceSelectorHelper<any AuthPaymentsService> {
    public static var payments: Self { .init() }
}

File: node/ts/net/chat/AuthPaymentsService.ts
//
// Copyright 2026 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

import { RequestOptions, AuthenticatedChatConnection } from '../Chat.js';
import * as NativeNice from '../../NativeNice.js';
import type { StandardNetworkError } from '../../Errors.js';
import type { Timestamp } from '../../NiceConverters.js';

declare module '../Chat' {
  // eslint-disable-next-line @typescript-eslint/no-empty-object-type
  interface AuthenticatedChatConnection extends AuthPaymentsService {}
}

export type CurrencyConversions = {
  timestampMs: Timestamp;
  currencies: Array<Currency>;
};

export type Currency = {
  base: string;
  /**
   * The values of this map are decimal conversion rates.
   */
  conversions: Map<string, string>;
};

export interface AuthPaymentsService {
  /**
   * Return the current currency conversion rates.
   *
   * @throws {StandardNetworkError}
   */
  getCurrencyConversions: (
    options?: RequestOptions
  ) => Promise<CurrencyConversions>;
}
AuthenticatedChatConnection.prototype.getCurrencyConversions = async function (
  options?: RequestOptions
): Promise<CurrencyConversions> {
  const { timestampMs, currencies } =
    await NativeNice.AuthenticatedChatConnection_get_currency_conversions({
      asyncContext: this.asyncContext,
      abortSignal: options?.abortSignal,
      chat: this.chatService,
    });
  return {
    timestampMs,
    currencies: currencies.map(({ base, conversions }) => ({
      base,
      conversions: new Map(conversions),
    })),
  };
};
